### PR TITLE
Method no wrapper

### PR DESCRIPTION
**Pull request by AGaliciaMartinez**
This is PR includes the main class Network along with the specialisations that handle the conversion between data types and matmul. It also includes a testing function to allow comparison between networks. Although this function is not in its final form, I will leave it as it is for now. 

A few differences with respect to QuOperator:

`Network` class is very similar to `QuOperator` in its logic. The main differences are:
 - We keep track of nodes. `QuOperator` navigates the graph starting from the in_edges and out_edges instead. I believe keeping track of the node reference can be useful, specially for larger tensor networks as it will significantly improve the speed on those cases. In any case, it is straightforward to keep track of nodes in most of the linear algebra operations.
- `ref_nodes` has been removed in favour of just `nodes`.
- `ignore_edges` attribute has been removed to simplify the implementation: I am not sure of the use case for this attribute but for the moment I decided to remove it just to simplify tests and `Network`.

Methods that work (if the have a [ ] is because there is some work still to do):
- `dims`
- `_check_*`: these checks are a little bit more exhausting that the ones in the original implementation of `QuOperator`.
- `_fast_constructor`
- `conj`, `transpose` and `adjoint`. These follow same logic as `QuOperator`. 
- `contract`: Same logic as `QuOperator`. It does not handle the Copy (identity) nodes yet. I will add them this feature in another PR.
- `to_array`: replaces `eval` in `QuOperator`.
- [x] `from_2d_array`: provisional. It replaces `from_tensor`. I am still not sure with the logic of this one. At this moment it handles 2D, 1d and scalar arrays so at least the name has to change.
- [x] `__matmul__`: The logic changes a bit as we relax the condition of equal dims to allow multiplication of different networks. They have to be compatible and this is checked with `_match_edges_by_split`. For example: we allow matmul between networks of dimension  `[[4],[4]]` and `[[2,2],[2,2]]`. Some more exaustive tests may be needed here.
- [x] `_match_edges_by_split`. This is the function that takes care of reshaping nodes in network to allow matmul. This funciton still requires some more careful testing but should work.
- [x] `tensor`: Not finished yet but I would like to add it in this PR.

Methods that will not be included in this PR:
- `partial_trace`
- `trace`
- `norm`
- `as_quoperator`: I still need google/TensorNetwork#951 to be merged.

A few other missing things:
- [ ] License
- [x] decide what to do with the `typing` module. Some functions have inherited the typing from `QuOperator`. Consder including it for every function or remove it.
- [x] Black/PEP8

### CODE REVIEW

File: src/qutip_tensornetwork/core/data/network.py [hunk 1]
```diff
@@ -0,0 +1,576 @@
...
+        # I need to check that edges are unique
+        self.out_edges = list(out_edges)
+        self.in_edges = list(in_edges)
+
+        # Unlike in QuOperator we will keep track of the nodes instead of
+        # dynamically searching for them when necessary. This is because
+        # searching all nodes in a large graph can be quite expensive while
+        # keeping track of them with network operations is straightforward.
+        self.nodes = set(nodes) if nodes else tn.reachable(self.in_edges
+                                                         + self.out_edges)
```
Comment (AGaliciaMartinez) [inline src/qutip_tensornetwork/core/data/network.py]: I am not sure how to store nodes. They need to be unique which is why in TensorNetwork they mostly use sets. However, this means that adding a new node will involve checking all of them to ensure the new node is unique. In most cases this is unnecessary. For instance, in matmul we copy both input networks so we now that the nodes are unique. I am thinking to change this into a list. Any thought @hodgestar, @nwlambert ?

File: src/qutip_tensornetwork/core/data/network.py [hunk 2]
```diff
@@ -0,0 +1,576 @@
...
+
+        self._check_edge_nodes_in_nodes()
+        self._check_in_out_are_dangling()
+        self._check_only_in_out_are_dangling()
+        self._check_edges_unique()
+        # I may need extra check such as for backends etc in the future
+
+        # TODO: should this be a porperty obtained from dims?
+        super().__init__(shape=(np.prod(self.dims[0]), np.prod(self.dims[1])))
```
Comment (AGaliciaMartinez) [inline src/qutip_tensornetwork/core/data/network.py]: At this moment the shape is an attribute. Given that the main objects are `in_edges` and `out_edges`, I wonder if we should define the `shape` as a property derived from those.

Comment (AGaliciaMartinez) [reply src/qutip_tensornetwork/core/data/network.py]: I decided to change it to an attribute.

File: src/qutip_tensornetwork/core/data/network.py [hunk 3]
```diff
@@ -0,0 +1,627 @@
...
+            A network that represents the conjugation of the input.
+        """
+        node_dict, edge_dict = tn.copy(self.nodes, conjugate=True)
+        nodes = set(node_dict[n] for n in self.nodes)
+        in_edges = [edge_dict[e] for e in self.in_edges]
+        out_edges = [edge_dict[e] for e in self.out_edges]
+
+        return Network._fast_constructor(out_edges, in_edges, nodes)
+
+    def transopose(self):
```
Comment (hodgestar) [inline src/qutip_tensornetwork/core/data/network.py]:
```suggestion
    def transpose(self):
```

File: src/qutip_tensornetwork/core/data/matmul.py [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+    defaulting to 1. If `out` is not given it is assumed to be 0.
+
+    Notes
+    -----
+    out as parameter is not included yet.
+    """
+    # This part of the operation wont be added until we think about what to do
+    # with addition.
+    if out is not None:
+        raise NotImplementedError()
```
Comment (hodgestar) [inline src/qutip_tensornetwork/core/data/matmul.py]:
```suggestion
        raise NotImplementedError("The addition of Network data layer instances is not yet implemented")
```

File: src/qutip_tensornetwork/core/data/matmul.py [hunk 5]
```diff
@@ -0,0 +1,30 @@
...
+
+    if scale != 1:
+        out._nodes = out._nodes | tn.Node(scale)
+
+    return left@right
```
Comment (hodgestar) [inline src/qutip_tensornetwork/core/data/matmul.py]:
```suggestion
    return left @ right
```

File: src/qutip_tensornetwork/testing.py [hunk 6]
```diff
@@ -0,0 +1,100 @@
...
+
+    rtol: float, optional
+        Absolute tolerance.
+
+    Raises
+    ------
+    AssertionError
+        If actual and desired do not represent the same map.
+    ValueError
+        If more than one node in the network is close to each other.
```
Comment (hodgestar) [inline src/qutip_tensornetwork/testing.py]:
```suggestion
```
This function is named `assert_network_close` so it should only raise AssertionErrors. It's okay if some of the helper functions it calls raise other exceptions, but this function should probably catch those and convert them into AssertionErrors.

File: src/qutip_tensornetwork/testing.py [hunk 7]
```diff
@@ -0,0 +1,100 @@
...
+    # We now compare Edges to see if they are properly connected.
+    for node in actual.nodes:
+        for edge in node.edges:
+            if not edge.is_dangling():
+                # This is a very basic test. We need to improve it to check
+                # whether the two edges are really the same: same dimentions
+                # and connected to same index.
+                assert_no_edge_between(node_dict[edge.node1], node_dict[edge.node2])
+
+def assert_no_edge_between(node1, node2):
```
Comment (hodgestar) [inline src/qutip_tensornetwork/testing.py]:
```suggestion
def assert_edge_between(node1, node2):
```
This function should be called `assert_edge_between` because it asserts that there *is* an edge shared between node1 and node2?

File: src/qutip_tensornetwork/testing.py [hunk 8]
```diff
@@ -0,0 +1,100 @@
+import numpy as np
+def assert_network_close(actual, desired, atol=0, rtol=1e-7):
```
Comment (hodgestar) [inline src/qutip_tensornetwork/testing.py:4]: Not necessarily for this PR, but graph isomorphism is in general a complex problem, so I recommend that we hit it with an appropriate hammer -- e.g. the networkx library -- in testing.

Rough plan:

* Add networkx as a testing dependency.
* Convert actual and desired to networkx graphs (maybe this is the one we want https://networkx.org/documentation/stable/reference/classes/multidigraph.html ?)
* Use is_isomorphic with custom node_match and edge_match arguments to check for isomorphism (see https://networkx.org/documentation/stable/reference/algorithms/generated/networkx.algorithms.isomorphism.is_isomorphic.html#networkx.algorithms.isomorphism.is_isomorphic).
* Assert that the graphs are isomorphic.

Perhaps such an equality check is worth exposing publicly too, but let's try it out in tests first.

File: tests/core/data/test_convert.py [hunk 9]
```diff
@@ -0,0 +1,44 @@
+"""This module tests that the basic conversion and creation methods in qutip
+work properly with networks."""
+import numpy as np
+import pytest
+import qutip
+from qutip_tensornetwork.core.data import Network
+from qutip import data
+from .conftest import random_complex_network, random_numpy_dense
+
+_network = random_complex_network(2)
```
Comment (hodgestar) [inline tests/core/data/test_convert.py]: Is there a good reason to create on random network here instead of just creating one in the functions that need them? Global mutable state can result in subtle bugs, so it's best avoided where possible.

File: tests/core/data/test_convert.py [hunk 10]
```diff
@@ -0,0 +1,44 @@
...
+                         ["network", Network],
+                         ids=["string `network`", "data type"])
+def test_convert_dense_to_network(shape, dtype):
+    array = random_numpy_dense(shape)
+    qobj = qutip.Qobj(array)
+    qobj = qobj.to(dtype)
+    assert isinstance(qobj.data, Network)
+    np.testing.assert_allclose(qobj.full(), array)
+
+def test_network_to_dense():
```
Comment (hodgestar) [inline tests/core/data/test_convert.py]: Perhaps we should test a few different shapes here too, since it's fairly easy to add and might catch problems with edge cases later.

File: tests/core/data/test_network.py [hunk 11]
```diff
@@ -0,0 +1,334 @@
...
+from qutip_tensornetwork.core.data import Network
+from qutip_tensornetwork.testing import assert_network_close
+from .conftest import random_node, random_complex_network
+import tensornetwork as tn
+
+def test_init_only_in_edge():
+    node = random_node((2,))
+    network = Network([], node[0:], copy=False)
+    assert len(network.nodes) == 1
+    assert network.nodes.pop() is node
```
Comment (hodgestar) [inline tests/core/data/test_network.py]:
```suggestion
    assert network.nodes[0] is node
```
Modify the `.nodes` while checking them seems a bit odd, so let's just do `.nodes[0]` instead if we can.

File: tests/core/data/test_network.py [hunk 12]
```diff
@@ -0,0 +1,334 @@
...
+    with pytest.raises(ValueError):
+       network = Network([node[0]], [node[0]])
+
+def test_non_dangling_in_or_out_raises():
+    """Input edges for in_edges and out_edges must be dangling."""
+    node = random_node((2,))
+    node2 = random_node((2,))
+    node[0] ^ node2[0]
+
+    with pytest.raises(ValueError):
```
Comment (hodgestar) [inline tests/core/data/test_network.py]: Could we also test that these raise suitable error messages?

File: tests/core/data/test_network.py [hunk 13]
```diff
@@ -0,0 +1,334 @@
...
+    """Included nodes can not have dangling edges that are not in_edges or
+    out_edges
+    """
+    node = random_node((2,2,2))
+    with pytest.raises(ValueError):
+        network = Network([node[0]], [node[1]], copy=False)
+
+def test_scalar_nodes():
+    """It is possible to include scalar nodes as long as they are not the only
+    nodes.
```
Comment (hodgestar) [inline tests/core/data/test_network.py]: This comment is a bit confusing. It says "as long as they are not the only nodes" but test_scalar_network successfully builds a network with only a scalar node. Perhaps the wording just needs to change a bit?

File: tests/core/data/test_network.py [hunk 14]
```diff
@@ -0,0 +1,334 @@
...
+    node = random_node(())
+    network = Network([], [], nodes=[node])
+    np.testing.assert_allclose(node.tensor, network.nodes.pop().tensor)
+
+def test_scalar_network_contraction():
+    node = random_node(())
+    network = Network([], [], nodes=[node])
+    np.testing.assert_allclose(node.tensor, network.to_array())
+
+def test_empty_rises():
```
Comment (hodgestar) [inline tests/core/data/test_network.py]:
```suggestion
def test_empty_raises():
```

File: tests/core/data/test_network.py [hunk 15]
```diff
@@ -0,0 +1,334 @@
...
+
+    copied_network = network.copy()
+    assert copied_network.in_edges[0] is not network.in_edges[0]
+    assert copied_network.out_edges[0] is not network.out_edges[0]
+
+    assert copied_network.nodes is not network.nodes
+    node = network.nodes.pop()
+    copied_node = copied_network.nodes.pop()
+    assert node is not copied_node
+    assert node.tensor is copied_node.tensor
```
Comment (hodgestar) [inline tests/core/data/test_network.py:120]: We should assert that the copied networks are equal, even if it's just with `network.to_array()` and `copied_network.to_array()` for now.

File: tests/core/data/test_network.py [hunk 16]
```diff
@@ -0,0 +1,334 @@
...
+    node = random_node((2,3))
+    network = Network([node[0]], [node[1]])
+
+    assert network.shape[0] == network.transopose().shape[1]
+    assert network.shape[1] == network.transopose().shape[0]
+    assert network.dims[0] == network.transopose().dims[1]
+    assert network.dims[1] == network.transopose().dims[0]
+
+    np.testing.assert_almost_equal(node.tensor,
+                                   network.transopose().nodes.pop().tensor)
```
Comment (hodgestar) [inline tests/core/data/test_network.py]:
```suggestion
    assert network.shape[0] == network.transpose().shape[1]
    assert network.shape[1] == network.transpose().shape[0]
    assert network.dims[0] == network.transpose().dims[1]
    assert network.dims[1] == network.transpose().dims[0]

    np.testing.assert_almost_equal(node.tensor,
                                   network.transpose().nodes.pop().tensor)
```

File: tests/core/data/test_network.py [hunk 17]
```diff
@@ -0,0 +1,334 @@
...
+
+class TestContract():
```
Comment (hodgestar) [inline tests/core/data/test_network.py]: I like the use of the `class TestXXX` here to group things, but I think we should make groups consistently -- so definitely we should add a `class TestNetwork` (for tests for methods of the Network class) and then one group for each specialization. Tests outside that can either be loose or grouped depending on what seems to make sense.

File: tests/core/data/test_network.py [hunk 18]
```diff
@@ -0,0 +1,334 @@
...
+    node1 = random_node((2,))
+    node2 = random_node((2, 2, 2))
+    node1[0] ^ node2[2]
+
+    network = Network([node2[0]], [node2[1]], [node1, node2])
```
Comment (hodgestar) [inline tests/core/data/test_network.py]: We should also just create this network within the individual tests if we can.

File: tests/core/data/test_network.py [hunk 19]
```diff
@@ -0,0 +1,334 @@
...
+class TestMatmul:
+
+    @pytest.mark.parametrize("dim_in, dim_out",
+                            [([4], [2,2]),
+                             ([2,2], [4]),
+                             ([2,10], [2,5,2]),
+                             ([2,10], [2,2,5]),
+                             ([2], [2]),
+                            ])
+    def test_comatible(self, dim_in, dim_out):
```
Comment (hodgestar) [inline tests/core/data/test_network.py]:
```suggestion
    def test_compatible(self, dim_in, dim_out):
```

File: tests/core/data/test_network.py [hunk 20]
```diff
@@ -0,0 +1,334 @@
...
+        different dimensions. This test does not check for the correct tensor 
+        structure but rather that the output is numerically correct once
+        contracted."""
+        right = random_node(dim_out)
+        left = random_node(dim_in)
+
+        right_net = Network(right[:], [])
+        left_net = Network([], left[:])
+
+        result = left_net@right_net
```
Comment (hodgestar) [inline tests/core/data/test_network.py]:
```suggestion
        result = left_net @ right_net
```

File: tests/core/data/test_network.py [hunk 21]
```diff
@@ -0,0 +1,334 @@
...
+
+        # Desired
+        left = left.tensor.reshape((1, np.prod(dim_out)))
+        right = right.tensor.reshape((np.prod(dim_in), 1))
+        desired = left@right
```
Comment (hodgestar) [inline tests/core/data/test_network.py]:
```suggestion
        desired = left @ right
```

File: tests/core/data/test_network.py [hunk 22]
```diff
@@ -0,0 +1,334 @@
...
+    def test_numerically_correct(self):
+        network1 = random_complex_network(5)
+        network2 = network1.adjoint()
+
+        expected = network2.to_array()@network1.to_array()
+        np.testing.assert_allclose(expected, (network2@network1).to_array())
+
+    def test_graph_structure(self):
+        """The operation tested here can be respresented as a graph in the
+        forllowing way:
```
Comment (hodgestar) [inline tests/core/data/test_network.py]:
```suggestion
        following way:
```

File: tests/core/data/test_network.py [hunk 23]
```diff
@@ -0,0 +1,334 @@
...
+                         ((1, 2), [[], [2]]),
+                         ((2), [[2], []]),
+                         ((), [[], []]),
+                        ]
+                        )
+def test_from_2d_array(shape, expected_dims):
+    array = np.random.random(shape)
+    network = Network.from_2d_array(array)
+
+    np.testing.assert_allclose(array, network.to_array().reshape(shape))
```
Comment (hodgestar) [inline tests/core/data/test_network.py:351]: Why do we need to call `.reshape(shape)` here? I would expect the shape of the input array to be retained?

File: src/qutip_tensornetwork/core/data/network.py [hunk 24]
```diff
@@ -0,0 +1,627 @@
...
+
+class Network(qutip.core.data.Data):
+    """Represents arbitrary quantum objects as tensor networks.
+
+    Tensor networks will be composed of ``tensornetwork.Nodes`` and
+    ``tensornetwork.Edges``. Nodes represent n-dimensional arrays whose
+    dimensions are connected with edges to other arrays. An edge represents the
+    contraction that needs to be carried out between two nodes. The contraction
+    of nodes is done in a `lazy` way, which means that is only carried out when
+    necessary or explicitly required with the contract method.
```
Comment (hodgestar) [inline src/qutip_tensornetwork/core/data/network.py]:
```suggestion
    necessary or explicitly requested with the contract method.
```

File: src/qutip_tensornetwork/core/data/network.py [hunk 25]
```diff
@@ -0,0 +1,627 @@
...
+
+    def __init__(self, out_edges, in_edges, nodes = None, copy = True):
+        """Creates a new ``Network`` from a tensor network.
+        This encapsulates an existing tensor network, interpreting it as a linear
+        operator.
+
+        The network is checked for consistency: All dangling edges must either be
+        in `out_edges`, `in_edges`, or `ignore_edges`.
+
+        Parameters
```
Comment (hodgestar) [inline src/qutip_tensornetwork/core/data/network.py]: It's more common and useful to document the parameters for `__init__` in the class docstring (since from user code one calls `MyClass(x, y, z)` and not `MyClass.__init__(...)`.

File: src/qutip_tensornetwork/core/data/network.py [hunk 26]
```diff
@@ -0,0 +1,627 @@
...
+    of this class as a ``QuOperator``.
+    """
+
+    def __init__(self, out_edges, in_edges, nodes = None, copy = True):
+        """Creates a new ``Network`` from a tensor network.
+        This encapsulates an existing tensor network, interpreting it as a linear
+        operator.
+
+        The network is checked for consistency: All dangling edges must either be
+        in `out_edges`, `in_edges`, or `ignore_edges`.
```
Comment (hodgestar) [inline src/qutip_tensornetwork/core/data/network.py]:
```suggestion
        in `out_edges` or `in_edges`.
```

File: src/qutip_tensornetwork/core/data/network.py (hunk 1 above)
Comment (hodgestar) [reply src/qutip_tensornetwork/core/data/network.py]: I think a set is fine for now. Checking "x in some_set" is very fast in Python (similar to "x in some_dict").

File: src/qutip_tensornetwork/core/data/network.py [hunk 27]
```diff
@@ -0,0 +1,627 @@
...
+        in_space = [e.dimension for e in self.in_edges]
+        return [out_space, in_space]
+
+    def _check_in_out_are_dangling(self):
+        for (i, e) in enumerate(self.out_edges):
+          if not e.is_dangling():
+            raise ValueError("output edge {} is not dangling!".format(i))
+        for (i, e) in enumerate(self.in_edges):
+          if not e.is_dangling():
+            raise ValueError("input edge {} is not dangling!".format(i))
```
Comment (hodgestar) [inline src/qutip_tensornetwork/core/data/network.py]:
```suggestion
        for (i, e) in enumerate(self.out_edges):
            if not e.is_dangling():
                raise ValueError("output edge {} is not dangling!".format(i))
        for (i, e) in enumerate(self.in_edges):
            if not e.is_dangling():
                raise ValueError("input edge {} is not dangling!".format(i))
```

File: src/qutip_tensornetwork/core/data/network.py [hunk 28]
```diff
@@ -0,0 +1,627 @@
...
+        known_edges = set(self.in_edges + self.out_edges)
+        all_dangling_edges = get_subgraph_dangling(self.nodes)
+        if known_edges != all_dangling_edges:
+            unexpected_edges = all_dangling_edges.difference(known_edges)
+            raise ValueError("the network includes unexpected dangling edges."
+                            + str(unexpected_edges))
+
+    def _check_edges_unique(self):
+        """Check that in_edges and out_edges are unique."""
+        if (len(set(self.in_edges)) != len(self.in_edges) or
```
Comment (hodgestar) [inline src/qutip_tensornetwork/core/data/network.py]: Should `in_edges` and `out_edges` not be unique when combined and not just separately? I.e should `len(set(in_edges + out_edges)) == len(in_edges) + len(out_edges)`?

File: src/qutip_tensornetwork/core/data/network.py [hunk 29]
```diff
@@ -0,0 +1,627 @@
...
+            return Network([], [], nodes=[node])
+
+        else:
+            node = tn.Node(array)
+            return Network(node[0:1], node[1:])
+
+    def as_quoperator(self):
+        # TODO: This can not be implemented unless tensornetwork includes the 
+        # __init__ file for the quantum module.
+        raise NotImplementedError()
```
Comment (hodgestar) [inline src/qutip_tensornetwork/core/data/network.py]: Let's just delete this for now, I think -- I can't see a reason for us to make a quoperator right now.

File: src/qutip_tensornetwork/core/data/network.py [hunk 30]
```diff
@@ -0,0 +1,627 @@
...
+        new_nodes_other, new_edges_other = tn.copy(other.nodes, False)
+
+        in_edges = [new_edges_self[e] for e in self.in_edges]
+        out_edges = [new_edges_other[e] for e in other.out_edges]
+
+        out_edges, in_edges = _match_edges_by_split(out_edges, in_edges)
+
+        # connect edges to create network for the result
+        for (e_in, e_out) in zip(in_edges, out_edges):
+          _ = e_in ^ e_out
```
Comment (hodgestar) [inline src/qutip_tensornetwork/core/data/network.py]:
```suggestion
            _ = e_in ^ e_out
```

**Review by hodgestar — CHANGES_REQUESTED** [review]
Sorry this took awhile to get to.  I think this is a solid start. Let's fix the small things and merge it and we can sort out all the other pieces in follow up PRs.

Other comments:

* I think we should remove `src/qutip_tensornetwork/file.txt`?
* I think `version.py` should not be committed?
* I think it's worthwhile running black over the code in a separate PR (or even just at the end of this PR) in order to sort out the formatting inherited from the quoperator files.

File: src/qutip_tensornetwork/core/data/network.py (hunk 28 above)
Comment (AGaliciaMartinez) [reply src/qutip_tensornetwork/core/data/network.py]: Indeed, good catch. We do not want someone including the same edge to both in_edges and out_edges

File: src/qutip_tensornetwork/core/data/network.py (hunk 29 above)
Comment (AGaliciaMartinez) [reply src/qutip_tensornetwork/core/data/network.py]: I agree.

File: src/qutip_tensornetwork/core/data/tensor.py [hunk 31]
```diff
@@ -0,0 +1,16 @@
...
+from .network import Network
+
+
+__all__ = ["tensor_network"]
+
+def tensor_network(left, right):
+    """Returns a network obtained from the tensor operation between left and
+    right.
+
+    See also:
```
Comment (AGaliciaMartinez) [inline src/qutip_tensornetwork/core/data/tensor.py]: Reminder for myself. I think this is not properly documented.

File: src/qutip_tensornetwork/testing.py [hunk 32]
```diff
@@ -0,0 +1,100 @@
...
+    for node in actual.nodes:
+        for edge in node.edges:
+            if not edge.is_dangling():
+                # This is a very basic test. We need to improve it to check
+                # whether the two edges are really the same: same dimentions
+                # and connected to same index.
+                assert_no_edge_between(node_dict[edge.node1], node_dict[edge.node2])
+
+def assert_no_edge_between(node1, node2):
+    """Raise an AssertionError if the given nodes do not have an edge connectin
```
Comment (AGaliciaMartinez) [inline src/qutip_tensornetwork/testing.py]:
```suggestion
    """Raise an AssertionError if the given nodes do not have an edge connecting
```

File: tests/core/data/test_network.py (hunk 16 above)
Comment (AGaliciaMartinez) [reply tests/core/data/test_network.py]: Sorry for this one.

File: tests/core/data/test_network.py (hunk 21 above)
Comment (AGaliciaMartinez) [reply tests/core/data/test_network.py]: I will run black after committing all the suggestions.

File: tests/core/data/test_network.py (hunk 23 above)
Comment (AGaliciaMartinez) [reply tests/core/data/test_network.py:351]: `to_array` is a method required from `Data`. I believe it is meant to return always a 2D array. This assumption is mostly employed in the `testing_mathematics` although I realized no explicit documentation is added to this method in `Data`.

File: src/qutip_tensornetwork/testing.py (hunk 8 above)
Comment (AGaliciaMartinez) [reply src/qutip_tensornetwork/testing.py:4]: I agree that we should relly in other packages for testing isomorphism. I am not sure if networkx is the right tool though as it seems to be quite slow compared to other packages: https://graph-tool.skewed.de/performance.  I must admit that have not looked too much into the other packages either so it may be possible that they are not flexible enough. Lets keep this for another PR although I will add an issue to remember this needs to be changed.

Comment (hodgestar) [reply src/qutip_tensornetwork/testing.py:4]: I think the page you linked to says why networkx is a good fit for our case -- our graphs will be comparatively small and its pure Python. We should probably check that the isomorphism is fast for reasonably sized networks though -- e.g. maybe a few thousand nodes -- if we're going to use it outside of tests. Probably calling back into custom node_match and edge_match functions from C is slow and a pain though, so not sure if graph-tool will support that or perform well on it.

File: src/qutip_tensornetwork/testing.py (hunk 6 above)
Comment (AGaliciaMartinez) [reply src/qutip_tensornetwork/testing.py]: Currently the function raises `ValueError` because is not able to deal with networks that have more than one similar node. I will change the `ValueError` to `NotImplementedError` to make it clear that this is a missing feature but we should change to use the package you mentioned for this task.

File: src/qutip_tensornetwork/testing.py (hunk 7 above)
Comment (AGaliciaMartinez) [reply src/qutip_tensornetwork/testing.py]: No, it only asserts if the for loop ends, which happens only when there is no edge between the given nodes. I changed it in an attempt to make it more clear.

File: tests/core/data/test_network.py (hunk 11 above)
Comment (AGaliciaMartinez) [reply tests/core/data/test_network.py]: unfortunately sets are not subscriptable. This is another reason I was considering to store nodes as lists.

Comment (AGaliciaMartinez) [reply tests/core/data/test_network.py]: I removed the pop as much as I could but I left it for those cases where I thought changing it for `list(network.nodes)[0]` was not worth.

File: tests/core/data/test_network.py (hunk 17 above)
Comment (AGaliciaMartinez) [reply tests/core/data/test_network.py]: All the methods here are for the `Network` class so I would rather not add an extra class layer for Network. Also, some classes only have one tests so I prefer not to surround them with a class. I can group the methods that have multiple tests each in one class though.

File: tests/core/data/test_network.py (hunk 23 above)
Comment (AGaliciaMartinez) [reply tests/core/data/test_network.py:351]: The assumption that to_array always returns a 2D array is also employed for the scipy integrators.

File: src/qutip_tensornetwork/core/data/network.py [hunk 33]
```diff
@@ -0,0 +1,627 @@
...
+            Instance of Network with a single node representing array.
+
+        Examples
+        --------
+        >>> array = np.array((2, 2))
+        >>> net = Network.from_2d_array(array)
+        >>> net.dims
+        [[2], [2]]
+
+        One dimensional arrays are understoof as kets and cero dimensional ones
```
Comment (AGaliciaMartinez) [inline src/qutip_tensornetwork/core/data/network.py]:
```suggestion
        One dimensional arrays are understood as kets and zero dimensional ones
```

File: src/qutip_tensornetwork/testing.py [hunk 34]
```diff
@@ -0,0 +1,100 @@
...
+    Parameters:
+    -----------
+    actual : list of nodes
+        ``Network`` obtained.
+
+    desired : list of nodes
+        Desired ``Network``.
+
+    atol: float, optional
+        Relative tolerance.
```
Comment (AGaliciaMartinez) [inline src/qutip_tensornetwork/testing.py]:
```suggestion
        Absolute tolerance.
```

File: src/qutip_tensornetwork/testing.py [hunk 35]
```diff
@@ -0,0 +1,100 @@
...
+
+    rtol: float, optional
+        Absolute tolerance.
```
Comment (AGaliciaMartinez) [inline src/qutip_tensornetwork/testing.py]:
```suggestion
        Relative tolerance.
```

File: tests/core/data/test_network.py (hunk 13 above)
Comment (AGaliciaMartinez) [reply tests/core/data/test_network.py:75]: `test_scalar_network` builds the network with an scalar node, `node_scalar`, and a non-scalar node, `node`.

File: tests/core/data/test_network.py (hunk 18 above)
Comment (AGaliciaMartinez) [reply tests/core/data/test_network.py:194]: I included the node and network creation inside the methods.

File: tests/core/data/test_network.py (hunk 15 above)
Comment (AGaliciaMartinez) [reply tests/core/data/test_network.py:124]: I can definitely include that assertion but I do not think it is necessary here. Most of the networks in this test file have a single node and that is to facilitate these kind of comparisons. For instance, here, by doing `assert node.tensor is copied_node.tensor` we already check that both the copied and original nodes have the exact same information. In fact, they contain the _same_ tensor which means they save in memory when copying the nodes.

File: tests/core/data/test_network.py (hunk 13 above)
Comment (AGaliciaMartinez) [reply tests/core/data/test_network.py]: At first I was not going to allow the creation of scalar networks but now I do. This test should be removed.

File: tests/core/data/test_network.py (hunk 12 above)
Comment (AGaliciaMartinez) [reply tests/core/data/test_network.py]: I am actually againts testingt the error message itself. I do see why testing the error type raised is important. Including the error type in the tests ensures that in the future this same function will continue raising the same error. This is important because a user may catch this specific error as part of its program's logic and run into problems if the error is not `ValueError`.  However, I think the user is not supposed to use the string from the error as part of its program logic. It is meant to be a clarification of what happened and how to use the class. In this sense, I see the error message fulfilling the same role as the documentation, which is not included in the tests either.

File: src/qutip_tensornetwork/core/data/network.py [hunk 36]
```diff
@@ -0,0 +1,633 @@
...
+            if e_in.dimension % e_out.dimension != 0:
+                raise ValueError(
+                    "edges are not compatible. The dimensions of in_edges: "
+                    + str(in_dims)
+                    + " whereas for out_edges: "
+                    + str(out_dims)
+                )
+            else:
+                # new_shape=(e_out.dimension, e_in.dimension//e_out.dimension)
+                # new_e_in, e_in = tn.split_edge(e_in, shape=new_shape)
```
Comment (AGaliciaMartinez) [inline src/qutip_tensornetwork/core/data/network.py:601]: remove

File: tests/core/data/test_network.py (hunk 12 above)
Comment (hodgestar) [reply tests/core/data/test_network.py]: Error messages are not documentation though -- they're generated by the program and so can have bugs that are not immediately obvious. E.g. I've often had cases where `raise ValueError(f"{x} must be less than {y}")` should have been `raise ValueError(f"{y} must be less than {x}")` or the error message is just nonsense or not easily readable and tests are the only way to catch those.

Yes, code should not generally depend on the error string itself, but it's fine for tests to depend on the error string if the string is generated by the code base being tested. While code shouldn't depend on the error string, humans do when they're debugging.

There are also other cases where serious bugs can result from error messages -- e.g. `raise ValueError(f"password {p} is invalid")` which is a common security hole. Luckily we don't encounter these kinds of cases too often in QuTiP.

File: tests/core/data/test_network.py (hunk 17 above)
Comment (hodgestar) [reply tests/core/data/test_network.py]: Happy for you to make the call on how to group the tests nicely.

File: src/qutip_tensornetwork/core/data/convert.py [hunk 37]
```diff
@@ -0,0 +1,24 @@
+import qutip
+from .network import Network
+import tensornetwork as tn
+
+
+def from_quoperator(quoperator):
+    """Create a Network from a ``Quoperator``."""
+    return Network(quoperator.out_edges, quoperator.in_edges, quoperator.ref_nodes)
```
Comment (hodgestar) [inline src/qutip_tensornetwork/core/data/convert.py]: Do we still want to have `from_quoperator`?

**Review by hodgestar — APPROVED** [review]
Happy for this to be merged once the license is added, but also happy to do a last review after the license is added.

**Comment by AGaliciaMartinez** [comment]
@hodgestar I have just added the same license that can be found in the quantum.py file in TensorNetwork. Could you take a look and let me know if I added it correctly. I also included a small clarification stating that changes to the original file follow the projects main license. Is this the correct approach for this case?

File: src/qutip_tensornetwork/core/data/network.py [hunk 38]
```diff
@@ -0,0 +1,642 @@
...
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""This file has been derived from the quantum/quantum.py file in the
+google/TensorNetwork repository. The above license corresponds to the one found
+in the original file. The original file has been extensively modified and these
+modifications follow the same license as the rest of the project does."""
```
Comment (hodgestar) [inline src/qutip_tensornetwork/core/data/network.py]:
```suggestion
#
# Note:
#
# This file has been derived from quantum.py in the google/TensorNetwork GitHub repository,
# https://github.com/google/TensorNetwork/blob/9b48cdd98777e4d7f26cccf29a562afe61e51da4/
# tensornetwork/quantum/quantum.py.
#
# The license above is the one from the original file. The file has been extensively modified. In
# addition to the license above, the modifications are also provided under the license in
# LICENSE.txt.
```

Comment (hodgestar) [reply src/qutip_tensornetwork/core/data/network.py]: I suggested changing this to a comment and I modified the text a bit to add an explicit link to the source file and to explicitly mention where the other license is.

Otherwise looks good!

**Review by hodgestar — CHANGES_REQUESTED** [review]

**Comment by AGaliciaMartinez** [comment]
I accepted your change so I will merger once the tests have passed. :smile:

**Review by hodgestar — APPROVED** [review]
